sql/plan/ddl_event.go

```diff
@@ -0,0 +1,558 @@
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			eventDetails.Starts = eventCreationTime
+		}
+		if c.Ends != nil {
+			eventDetails.HasEnds = true
+			eventDetails.Ends, err = c.Ends.EvalTime(ctx)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	eventDetails.Created = eventCreationTime
+	eventDetails.LastAltered = eventCreationTime
+
+	return &createEventIter{
+		ed: eventDetails,
+		db: c.db,
+	}, nil
+}
+
+// createEventIter is the row iterator for *CreateEvent.
+type createEventIter struct {
+	once sync.Once
+	ed   sql.EventDetails
+	db   sql.Database
+}
+
+// Next implements the sql.RowIter interface.
+func (c *createEventIter) Next(ctx *sql.Context) (sql.Row, error) {
+	run := false
+	c.once.Do(func() {
+		run = true
+	})
+	if !run {
+		return nil, io.EOF
+	}
+
+	pdb, ok := c.db.(sql.EventDatabase)
+	if !ok {
+		return nil, sql.ErrEventsNotSupported.New(c.db.Name())
+	}
+
+	if c.ed.HasEnds {
+		if c.ed.Ends.Sub(c.ed.Starts).Seconds() < 0 {
+			return nil, fmt.Errorf("ENDS is either invalid or before STARTS")
+		}
+	}
+
+	err := pdb.SaveEvent(ctx, c.ed)
+	if err != nil {
+		return nil, err
+	}
+
+	if c.ed.HasExecuteAt {
+		// If the event execution time is in the past and ON COMPLETION NOT PRESERVE is set.
+		// The event is dropped immediately after creation.
+		if c.ed.ExecuteAt.Sub(c.ed.Created).Seconds() < 0 && !c.ed.OnCompletionPreserve {
+			err = pdb.DropEvent(ctx, c.ed.Name)
+			if err != nil {
+				return nil, err
+			}
+
+			ctx.Session.Warn(&sql.Warning{
+				Level:   "Note",
+				Code:    1588,
+				Message: fmt.Sprintf("Event execution time is in the past and ON COMPLETION NOT PRESERVE is set. The event was dropped immediately after creation."),
+			})
+			return sql.Row{types.NewOkResult(0)}, nil
+		}
+	}
+
+	// If Starts is set to current_timestamp or not set, then execute the event once and update last executed At.
+	if c.ed.Created.Sub(c.ed.Starts).Abs().Seconds() <= 1 {
+		// TODO: execute the event once and update 'LastExecuted' and 'ExecutionCount'
+	}
+
+	return sql.Row{types.NewOkResult(0)}, nil
+}
+
+// Close implements the sql.RowIter interface.
+func (c *createEventIter) Close(ctx *sql.Context) error {
+	return nil
+}
+
+// onScheduleEveryString returns ON SCHEDULE EVERY clause part of CREATE EVENT statement.
+func onScheduleEveryString(every sql.Expression, starts, ends *OnScheduleTimestamp) string {
+	everyInterval := strings.TrimPrefix(every.String(), "INTERVAL ")
+	startsStr := ""
+	if starts != nil {
+		startsStr = fmt.Sprintf(" STARTS %s", starts.String())
+	}
+	endsStr := ""
+	if ends != nil {
+		endsStr = fmt.Sprintf(" ENDS %s", ends.String())
+	}
+
+	return fmt.Sprintf("ON SCHEDULE EVERY %s%s%s", everyInterval, startsStr, endsStr)
+}
+
+// OnScheduleTimestamp is object used for EVENT ON SCHEDULE { AT / STARTS / ENDS } optional fields only.
+type OnScheduleTimestamp struct {
+	timestamp sql.Expression
+	intervals []sql.Expression
+}
+
+var _ sql.Node = (*OnScheduleTimestamp)(nil)
```

Comment:
Why is this defined as a Node? It can't be iterated (and isn't iterated from its parent object)

It should just be a `sql.Expression`, right? What I am I missing?